netdata/netdata
# Switch to using Debian as a base for our Docker images.

##### Summary

As we now have things which depend on libsystemd, and the systemd developers refuse to acknowledge anything other than glibc as a legitimate libc implementation for Linux, we are no longer able to provide full functionality in our Docker images using Alpine.

Most of the actual work for this is in https://github.com/netdata/helper-images/pull/223. This PR largely just tweaks a few things for correct compatibility behavior with the new base images.

Notable changes resulting from this switch:

- A handful of things that are included in our existing Docker images will no longer be included by default. This is due to package conflicts that are present on Debian but not Alpine. The primary things affected by this are NUT and the APCUPS tooling, both of which will need to be added at runtime if needed.
- Because of the above change together with a few extra bits of cleanup in the base images, the container images are actually about 10 MiB _smaller_ with this change.
- The existing infrastructure for adding extra packages at runtime has been updated to use a new variable name (it’s now `NETDATA_EXTRA_DEB_PACKAGES` instead of `NETDATA_EXTRA_APK_PACKAGES`). Users who are not utilizing this functionality will see no change. Users who have the old variable defined will see warnings in their container logs about this change, but their containers will still run (albeit without the additional packages specified in the old variable).

##### Test Plan

Docker CI is expected to fail on this PR until https://github.com/netdata/helper-images/pull/223 is merged and the images from that merge are published. That PR should not be merged until this one is otherwise ready to merge, as it will break the existing Docker image build process (specifically, the adduser, addgroup, and usermod changes in this PR are required to get the Docker image to build correctly with the updated base images).

Local testing requires the following steps to build the new container images:

1. Ensure that you _do not_ have local copies of the `netdata/base` or `netdata/builder` images on your build system.
2. Check out https://github.com/netdata/helper-images/pull/223.
3. Build a docker image from the above helper-images PR using the `base/Dockerfile` Dockerfile with the _exact_ image name `netdata/base:latest`. For example, from the root of the checked out helper-images repo: `docker build . -f base/Dockerfile -t netdata/base:latest`.
4. Build a docker image from the above helper-images PR using the `builder/Dockerfile` Dockerfile with the _exact_ image name `netdata/builder:latest`. For example, from the root of the checked out helper-images repo: `docker build . -f builder/Dockerfile -t netdata/builder:latest`.
5. Check out a copy of this PR locally.
6. Build a Docker image from this PR using our regular Dockerfile. Note that if you are using Podman instead of Docker, you must add the `--format=docker` option to the build command to get this image to build correctly.

##### Additional Information

Prompted by a desire to have the systemd-journal plugin working in our Docker containers.

Currently only build tested, I’ll be doing more rigorous testing tomorrow.
